xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/CoinmateStreamingServiceFactory.java

```diff
@@ -0,0 +1,27 @@
+package info.bitrich.xchangestream.coinmate;
+
+import info.bitrich.xchangestream.coinmate.CoinmateStreamingService;
+import info.bitrich.xchangestream.coinmate.dto.auth.AuthParams;
+
+public class CoinmateStreamingServiceFactory {
+
+  private AuthParams authParams;
+  private String baseUrl;
+
+  public CoinmateStreamingServiceFactory(String baseUrl, AuthParams authParams) {
+    this.baseUrl = baseUrl;
+    this.authParams = authParams;
+  }
+
+  public CoinmateStreamingService createAndConnect(String endpoint, boolean needsAuth) {
+    String url = baseUrl + "/" + endpoint;
+    if (needsAuth && authParams != null) {
+      url += "?" + authParams.toParams();
+    }
+    CoinmateStreamingService service = new CoinmateStreamingService(url);
+    // block until connected, because of nonce conflicts when connecting to multiple channels
+    service.connect().blockingAwait();
```

Comment:
Rather than blocking the calling thread, which breaks the RX guarantee, how about this:

- Use a `static` `Lock` to prevent concurrent connections
- Return `Completable<CoinmateStreamingService>` from this method (i.e. just return the result of `service.connect()`)

The, the parent service's `connect()` method will work asynchronously as expected.